File: app/src/main/java/no/sintef/fiskinfo/repository/SnapRepository.java
package no.sintef.fiskinfo.repository;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import java.util.ArrayList;
import java.util.List;

import no.sintef.fiskinfo.api.SnapMessageService;
import no.sintef.fiskinfo.model.EchogramInfo;
import no.sintef.fiskinfo.model.SnapMessage;
import no.sintef.fiskinfo.repository.dummy.DummyEchogram;
import no.sintef.fiskinfo.repository.dummy.DummySnap;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class SnapRepository {
    static SnapRepository instance;
    SnapMessageService snapMessageService = null;

    //    final static String SNAP_FISH_SERVER_URL = "https://10.218.86.229:44387/";
//    final static String SNAP_FISH_SERVER_URL = "http://10.218.69.173:58196/";
    final static String DEFAULT_SNAP_FISH_SERVER_URL = "http://10.218.86.229:5002/";

    String snapFishServerUrl = DEFAULT_SNAP_FISH_SERVER_URL;

    public SnapRepository(Context context) {
        updateFromPreferences(context);
    }

    public static SnapRepository getInstance(Context context) {
        if (instance == null)
            instance = new SnapRepository(context);
        return instance;
    }

    protected MutableLiveData<List<SnapMessage>> outboxSnaps;

    final MutableLiveData<List<SnapMessage>> inboxSnaps = new MutableLiveData<>();

    public LiveData<List<SnapMessage>> getInboxSnaps() {
        refreshInboxContent();
        return inboxSnaps;
    }



    protected void initService() {
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(snapFishServerUrl)
                .addConverterFactory(GsonConverterFactory.create())
                .build();

        snapMessageService = retrofit.create(SnapMessageService.class);

    }


    private void initOutbox() {
        outboxSnaps = new MutableLiveData<>();
        outboxSnaps.setValue(new ArrayList<SnapMessage>());
    }

    public void storeSnap(SnapMessage newSnap) {
        if (outboxSnaps == null)
            initOutbox();

        if (snapMessageService == null)
            initService();
        snapMessageService.sendSnapMessage(newSnap).enqueue(new Callback<SnapMessage>() {
            @Override
            public void onResponse(Call<SnapMessage> call, Response<SnapMessage> response) {

            }

            @Override
            public void onFailure(Call<SnapMessage> call, Throwable t) {

            }
        });
        outboxSnaps.getValue().add(newSnap);
    }

    public LiveData<List<SnapMessage>> getOutboxSnaps() {
        if (outboxSnaps == null) {
            outboxSnaps = new MutableLiveData<>();
            outboxSnaps.setValue(new ArrayList<SnapMessage>());
        }
        return outboxSnaps;
    }

    public LiveData<List<EchogramInfo>> getEchogramInfos() {
        if (snapMessageService == null)
            initService();


        final MutableLiveData<List<EchogramInfo>> data = new MutableLiveData<>();

        // This implementation is still suboptimal but better than before.
        // A complete implementation also handles error cases.
        snapMessageService.getEchogramInfos().enqueue(new Callback<List<EchogramInfo>>() {
            @Override
            public void onResponse(Call<List<EchogramInfo>> call, Response<List<EchogramInfo>> response) {
                data.setValue(response.body());
            }

            @Override
            public void onFailure(Call<List<EchogramInfo>> call, Throwable t) {
                data.setValue(DummyEchogram.getDummyEchograms());
            }
        });
        return data;
    }

    public void refreshInboxContent() {
        if (snapMessageService == null)
            initService();

        snapMessageService.getSnapMessages(true).enqueue(new Callback<List<SnapMessage>>() {
            @Override
            public void onResponse(Call<List<SnapMessage>> call, Response<List<SnapMessage>> response) {
                inboxSnaps.setValue(response.body());
            }

            @Override
            public void onFailure(Call<List<SnapMessage>> call, Throwable t) {
                inboxSnaps.setValue(DummySnap.getDummyInboxSnaps().getValue());
            }
        });
    }


    public void updateFromPreferences(Context context) {
        if (context != null) {
            // Find preferences
            SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
            snapFishServerUrl = prefs.getString("server_address", DEFAULT_SNAP_FISH_SERVER_URL);
            snapMessageService = null;
        }
        refreshInboxContent();
    }

}

File: app/src/main/java/no/sintef/fiskinfo/repository/SnapRepository.kt
/**
 * Copyright (C) 2020 SINTEF
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package no.sintef.fiskinfo.repository

import android.content.Context

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.preference.PreferenceManager
import no.sintef.fiskinfo.R

import java.util.ArrayList

import no.sintef.fiskinfo.api.snapfish.SnapMessageService
import no.sintef.fiskinfo.api.createService
import no.sintef.fiskinfo.model.SnapMetadata
import no.sintef.fiskinfo.model.SnapMessage
import no.sintef.fiskinfo.model.SnapMessageDraft
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class SnapRepository(context: Context) {
    internal var snapMessageService: SnapMessageService? = null

    internal var snapFishServerUrl: String? = DEFAULT_SNAP_FISH_SERVER_URL

    internal var snapFishUserId: Int = 2

    internal var outboxSnaps = MutableLiveData<List<SnapMessage>>()
    internal val inboxSnaps = MutableLiveData<List<SnapMessage>>()

    internal val echogramInfos = MutableLiveData<List<SnapMetadata>>()

    init {
        updateFromPreferences(context)
    }

    fun getInboxSnaps(): LiveData<List<SnapMessage>> {
        refreshInboxContent()
        return inboxSnaps
    }

    private fun initService() {
        snapMessageService = createService(SnapMessageService::class.java, snapFishServerUrl!!, true)
    }

/*
    protected fun initService() {
        val retrofit = Retrofit.Builder()
            .baseUrl(snapFishServerUrl!!)
            .addConverterFactory(GsonConverterFactory.create())
            .build()

        snapMessageService = retrofit.create(SnapMessageService::class.java)

    }
*/

    private fun initOutbox() {
        if (outboxSnaps.value == null)
            outboxSnaps.value = ArrayList()
    }

    fun sendSnap(newSnap: SnapMessageDraft) {
        if (outboxSnaps.value == null)
            initOutbox()

        if (snapMessageService == null)
            initService()
        snapMessageService!!.sendSnapMessage(newSnap).enqueue(object : Callback<SnapMessage> {
            override fun onResponse(call: Call<SnapMessage>, response: Response<SnapMessage>) {
                if (response.body() != null) {
                    (outboxSnaps.value!! as ArrayList<SnapMessage>).add(response.body()!!)
                    (outboxSnaps.value!! as ArrayList<SnapMessage>).sortByDescending { it.sentTimestamp }
                }
            }

            override fun onFailure(call: Call<SnapMessage>, t: Throwable) {

            }
        })
    }

    fun deleteInboxSnap(message: SnapMessage) {
        inboxSnaps.value = inboxSnaps.value!!.minusElement(message)
    }

    fun deleteOutboxSnap(message: SnapMessage) {
        outboxSnaps.value = outboxSnaps.value!!.minusElement(message)
    }

    fun refreshOutboxContent() {
        if (outboxSnaps.value == null)
            initOutbox()
    }

    fun getOutboxSnaps(): LiveData<List<SnapMessage>> {
        if (outboxSnaps.value == null) {
            initOutbox()
        }
        outboxSnaps.value = outboxSnaps.value?.sortedByDescending {it.sentTimestamp} ?: ArrayList()

        return outboxSnaps
    }

    fun getEchogramInfos(): LiveData<List<SnapMetadata>> {
        refreshEchogramListContent()
        return echogramInfos
    }

    fun refreshInboxContent() {
        if (snapMessageService == null)
            initService()

        snapMessageService!!.getSnapMessages(snapFishUserId, inbox = true, snapmetadata = true).enqueue(object : Callback<List<SnapMessage>> {
            override fun onResponse(call: Call<List<SnapMessage>>, response: Response<List<SnapMessage>>) {
                inboxSnaps.value = response.body()?.sortedByDescending {it.sentTimestamp} ?: ArrayList()
            }

            override fun onFailure(call: Call<List<SnapMessage>>, t: Throwable) {
                // TODO: log problem
                inboxSnaps.value = ArrayList()
            }
        })
    }

    fun refreshEchogramListContent() {
        if (snapMessageService == null)
            initService()

        snapMessageService!!.getSnapMetadata(snapFishUserId).enqueue(object : Callback<List<SnapMetadata>> {
            override fun onResponse(call: Call<List<SnapMetadata>>, response: Response<List<SnapMetadata>>) {
                echogramInfos.value = response.body()?.sortedByDescending {it.timestamp} ?: ArrayList()
            }

            override fun onFailure(call: Call<List<SnapMetadata>>, t: Throwable) {
                // TODO: log problem
                echogramInfos.value = ArrayList()
            }
        })
    }


    fun updateFromPreferences(context: Context?) {
        if (context != null) {
            // Find preferences
            val prefs = PreferenceManager.getDefaultSharedPreferences(context)
            snapFishServerUrl = prefs.getString( context.getString(R.string.pref_snap_api_server_address), DEFAULT_SNAP_FISH_SERVER_URL)
            snapFishUserId = prefs.getString( context.getString(R.string.pref_user_id),"2")!!.toInt()
            snapMessageService = null
        }
        refreshInboxContent()
    }

    companion object {
        var instance: SnapRepository? = null

        internal const val DEFAULT_SNAP_FISH_SERVER_URL = "http://129.242.16.123:37789/"
        fun getInstance(context: Context): SnapRepository {
            if (instance == null)
                instance = SnapRepository(context)
            return instance!!
        }
    }

}
